File: kotlin/src/commonMain/kotlin/org/meshtastic/tak/DictionaryProvider.kt
package org.meshtastic.tak

/**
 * Loads and provides zstd compression dictionaries.
 *
 * The raw dictionary bytes are fetched through the internal [DictionaryLoader]
 * (a single commonMain object backed by the generated `EmbeddedDictionaries`,
 * uniform across every target as of v0.6.0). This provider owns the canonical
 * resource names, the dictionary-ID constants, and the aircraft-vs-non-aircraft
 * selection rule.
 */
public object DictionaryProvider {
    /**
     * Dictionary ID for the 512 KB non-aircraft, proto-trained dictionary —
     * the default for PLI, chat, ground units, shapes, markers, routes, etc.
     * Encoded in flags-byte bits 0–5 of the wire payload.
     */
    public const val DICT_ID_NON_AIRCRAFT: Int = 0

    /**
     * Dictionary ID for the 4 KB aircraft dictionary, used for Air-domain CoT
     * types (3rd type atom = `A`). Encoded in flags-byte bits 0–5.
     */
    public const val DICT_ID_AIRCRAFT: Int = 1

    /**
     * Reserved flags-byte value (`0xFF`) meaning the payload is raw,
     * uncompressed `TAKPacketV2` protobuf — no zstd, no dictionary. Emitted by
     * the encoder's skip-compress path and by TAK_TRACKER firmware.
     */
    public const val DICT_ID_UNCOMPRESSED: Int = 0xFF

    /** Lazily-loaded bytes of the non-aircraft dictionary ([DICT_ID_NON_AIRCRAFT]). */
    public val nonAircraftDict: ByteArray by lazy {
        DictionaryLoader.loadDictionary("dict_non_aircraft.zstd")
    }

    /** Lazily-loaded bytes of the aircraft dictionary ([DICT_ID_AIRCRAFT]). */
    public val aircraftDict: ByteArray by lazy {
        DictionaryLoader.loadDictionary("dict_aircraft.zstd")
    }

    /**
     * Return the dictionary bytes for a dictionary ID.
     *
     * @param dictId [DICT_ID_NON_AIRCRAFT] or [DICT_ID_AIRCRAFT].
     * @return the dictionary bytes, or `null` for [DICT_ID_UNCOMPRESSED] and any
     *         unknown ID (the decoder treats a `null` here as "unknown
     *         dictionary" and rejects the frame).
     */
    public fun getDictionary(dictId: Int): ByteArray? =
        when (dictId) {
            DICT_ID_NON_AIRCRAFT -> nonAircraftDict
            DICT_ID_AIRCRAFT -> aircraftDict
            else -> null
        }

    /**
     * Pick the dictionary ID to compress a packet with, based on its CoT type.
     *
     * Prefers the enum classification ([CotTypeMapper.isAircraft]); when the
     * type is [CotTypeMapper.COTTYPE_OTHER] it falls back to the raw type string
     * ([CotTypeMapper.isAircraftString]). Air-domain types map to
     * [DICT_ID_AIRCRAFT], everything else to [DICT_ID_NON_AIRCRAFT].
     *
     * @param cotTypeId the packet's `cot_type_id` enum value.
     * @param cotTypeStr the raw CoT type string, used only when [cotTypeId] is
     *        [CotTypeMapper.COTTYPE_OTHER]; may be `null`.
     * @return [DICT_ID_AIRCRAFT] or [DICT_ID_NON_AIRCRAFT].
     */
    public fun selectDictId(
        cotTypeId: Int,
        cotTypeStr: String?,
    ): Int {
        // Check enum-based classification first
        if (cotTypeId != CotTypeMapper.COTTYPE_OTHER) {
            return if (CotTypeMapper.isAircraft(cotTypeId)) DICT_ID_AIRCRAFT else DICT_ID_NON_AIRCRAFT
        }
        // Fall back to string-based classification
        if (cotTypeStr != null && CotTypeMapper.isAircraftString(cotTypeStr)) {
            return DICT_ID_AIRCRAFT
        }
        return DICT_ID_NON_AIRCRAFT
    }
}

File: typescript/src/DictionaryProvider.ts
/**
 * Loads the canonical zstd dictionaries from the package's `resources/`
 * directory and selects the right dictionary for a given CoT type.
 *
 * Two dictionaries ship with the SDK, both pre-trained on serialized
 * `TAKPacketV2` protobuf bytes (not raw XML): a ~512 KB non-aircraft dictionary
 * and a ~4 KB aircraft dictionary. The dictionary ID also doubles as the wire
 * `flags` byte value (bits 0–5), so the same constants name both the dictionary
 * and the on-wire flag. All nodes on a mesh must ship the same dictionary pair
 * to interoperate; the dictionaries are static and never adapted from runtime
 * traffic, which keeps every packet independently decodable.
 *
 * @remarks
 * Dictionary buffers are read from disk lazily and cached for the process
 * lifetime — the first call to {@link nonAircraftDict} / {@link aircraftDict}
 * pays the file read, subsequent calls are free.
 */
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";
import { isAircraft, isAircraftString, COTTYPE_OTHER } from "./CotTypeMapper.js";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const RESOURCES_DIR = path.resolve(__dirname, "../resources");

/**
 * Dictionary ID / flags-byte value for the non-aircraft (~512 KB) dictionary.
 * Covers PLI, GeoChat, ground units, shapes, markers, routes, ranging, alerts,
 * casevac, emergency, task, TAKTALK, and delete events.
 */
export const DICT_ID_NON_AIRCRAFT = 0;

/**
 * Dictionary ID / flags-byte value for the aircraft (~4 KB) dictionary.
 * Covers ADS-B tracks, military air, and helicopters.
 */
export const DICT_ID_AIRCRAFT = 1;

/**
 * Sentinel flags-byte value (`0xFF`) marking an uncompressed payload: the body
 * is raw `TAKPacketV2` protobuf with no zstd compression. Emitted by the
 * encoder's skip-compress path when compression wouldn't shrink the payload,
 * and by firmware devices that cannot compress.
 */
export const DICT_ID_UNCOMPRESSED = 0xff;

let _nonAircraftDict: Buffer | null = null;
let _aircraftDict: Buffer | null = null;

/**
 * Load (and cache) the non-aircraft zstd dictionary from `resources/`.
 *
 * @returns The raw dictionary bytes.
 * @throws If `resources/dict_non_aircraft.zstd` cannot be read.
 */
export function nonAircraftDict(): Buffer {
  if (!_nonAircraftDict) {
    _nonAircraftDict = fs.readFileSync(path.join(RESOURCES_DIR, "dict_non_aircraft.zstd"));
  }
  return _nonAircraftDict;
}

/**
 * Load (and cache) the aircraft zstd dictionary from `resources/`.
 *
 * @returns The raw dictionary bytes.
 * @throws If `resources/dict_aircraft.zstd` cannot be read.
 */
export function aircraftDict(): Buffer {
  if (!_aircraftDict) {
    _aircraftDict = fs.readFileSync(path.join(RESOURCES_DIR, "dict_aircraft.zstd"));
  }
  return _aircraftDict;
}

/**
 * Resolve a dictionary ID (from a received flags byte, masked with `& 0x3F`)
 * to its dictionary bytes.
 *
 * @param dictId - {@link DICT_ID_NON_AIRCRAFT} or {@link DICT_ID_AIRCRAFT}.
 * @returns The matching dictionary buffer, or `null` for any unknown ID
 *          (including {@link DICT_ID_UNCOMPRESSED}, which has no dictionary).
 */
export function getDictionary(dictId: number): Buffer | null {
  if (dictId === DICT_ID_NON_AIRCRAFT) return nonAircraftDict();
  if (dictId === DICT_ID_AIRCRAFT) return aircraftDict();
  return null;
}

/**
 * Choose the dictionary ID to compress a packet with, based on its CoT type.
 *
 * Aircraft types use {@link DICT_ID_AIRCRAFT}; everything else uses
 * {@link DICT_ID_NON_AIRCRAFT}. When the type is known
 * ({@link COTTYPE_OTHER} not returned by the enum) the decision is made from
 * the enum via {@link isAircraft}; otherwise it falls back to classifying the
 * raw `cotTypeStr` with {@link isAircraftString}.
 *
 * @param cotTypeId  - The packet's `CotType` enum value.
 * @param cotTypeStr - The raw CoT type string, used only when `cotTypeId` is
 *                     {@link COTTYPE_OTHER}.
 * @returns {@link DICT_ID_AIRCRAFT} or {@link DICT_ID_NON_AIRCRAFT}.
 */
export function selectDictId(cotTypeId: number, cotTypeStr?: string): number {
  if (cotTypeId !== COTTYPE_OTHER) {
    return isAircraft(cotTypeId) ? DICT_ID_AIRCRAFT : DICT_ID_NON_AIRCRAFT;
  }
  if (cotTypeStr && isAircraftString(cotTypeStr)) return DICT_ID_AIRCRAFT;
  return DICT_ID_NON_AIRCRAFT;
}
